web-backend/src/main/java/com/statusneo/vms/service/VisitService.java

```diff
@@ -103,4 +104,23 @@ public Visit registerVisit(Visitor visitor) {
         return savedVisit;
     }
 
+    @Transactional
+    public VerificationResult confirmVisit(Long visitId, String otpCode) {
+        Visit visit = visitRepository.findById(visitId)
+                .orElseThrow(() -> new IllegalArgumentException("Visit not found for ID: " + visitId));
+
+        boolean isValid = otpService.validateOtp(visit, otpCode);
+
+        if (!isValid) {
+            boolean shouldReattempt = !otpService.hasExceededOtpAttempts(visit);
+            return new VerificationResult(false, shouldReattempt, shouldReattempt
+                    ? "Invalid OTP. Please try again." : "Maximum attempts exceeded.");
+        }
+
+        visit.setIsApproved(true);
+        visitRepository.save(visit);
+
+        return new VerificationResult(true, false, "OTP verified successfully");
```

Comment:
Good progress, but don't do this here. VerificationResult should be returned from OTPService.validateOtp(). We want to encapsulate all the logic of re-attempts there. So hasExceededOtpAttempts() doesn't need to be exposed outside, it's all internal. If we have to change no. of re-attempts, then it's just 1 config change inside OTPService and nothing changes here in VisitService.
Simply check the VerificationResult.success flag here.